### PR TITLE
Add support of external runtimes to exonum-cli

### PR DESCRIPTION
**Pull request by popzxc**
Currently we aren't able to create a node with additional runtimes with `exonum-cli`.

### CODE REVIEW

File: examples/cryptocurrency/examples/demo.rs [hunk 1]
```diff
@@ -64,10 +64,14 @@ fn node_config() -> NodeConfig {
 fn main() {
     exonum::helpers::init_logger().unwrap();
 
+    let external_runtimes: Vec<(u32, Box<dyn exonum::runtime::Runtime>)> = vec![];
```
Comment (popzxc) [inline examples/cryptocurrency/examples/demo.rs]: I think we should refine `Node` and `Blockchain` creation algorithm and enforce builders usage.

So we will have a much more cleaner code and encapsulate the `Node` internal structure:
```rust
let node = NodeBuilder::new()
    .with_db(TemporaryDB::new()
    .with_service(CryptocurrencyService)
    .finalize(node_config());
```

File: exonum/src/node/mod.rs [hunk 2]
```diff
@@ -909,6 +910,7 @@ impl Node {
     /// Creates node for the given services and node configuration.
     pub fn new(
         database: impl Into<Arc<dyn Database>>,
+        external_runtimes: impl IntoIterator<Item = impl Into<(u32, Box<dyn Runtime>)>>,
```
Comment (aleksuss) [inline exonum/src/node/mod.rs]: Just out of curiosity. Is there really reason to use such complicated type rather then `Vec<(u32, Box<dyn Runtime)>` or something like this ?

File: examples/cryptocurrency/examples/demo.rs (hunk 1 above)
Comment (aleksuss) [reply examples/cryptocurrency/examples/demo.rs]: Good idea. 👍🏻

File: exonum/src/node/mod.rs (hunk 2 above)
Comment (popzxc) [reply exonum/src/node/mod.rs]: Well, I thought about it too. We have even a much funnier place: function takes `IntoIter` just to `collect` it into a `vec`, and then (!) call `into_iter` for it to pass into another function: [src](https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/blockchain/builder.rs#L75).

However, I wanted this change to be consistent with surrounding code, so I made it `IntoIter` just as `services`.

I think it's a good topic for further refactoring, but fixing the whole issue is a bit out of scope of this PR.

Comment (aleksuss) [reply exonum/src/node/mod.rs]: Ok. Obviously it makes sense for now.

File: exonum/src/blockchain/builder.rs [hunk 3]
```diff
@@ -89,6 +88,17 @@ impl BlockchainBuilder {
         self.with_additional_runtime(runtime)
     }
 
+    pub fn with_external_runtimes(
```
Comment (vitvakatu) [inline exonum/src/blockchain/builder.rs]: Now we have 4 (!) methods for builder that are used to add Runtime:
`with_external_runtimes`, `with_additional_runtime`, `with_default_runtime` and `with_rust_runtime`. Don't you think it's somewhat complicated?

Comment (popzxc) [reply exonum/src/blockchain/builder.rs]: Yeah, it should be refined too.

**Comment by popzxc** [comment]
Regarding "this change will not bring any benefits yet":

1. I'm working on an additional runtime on my own, and this runtime doesn't require a complex initialization.
2. We have a `sample_runtime` example, which, I think should be refactored to use `exonum-cli`.

So it will bring some benefits.

**Comment by popzxc** [comment]
> Interesting will we have to allow adding of service instance collections directly in the `NodeBuilder` to create own networks with payload at startup.

Yes, I think it's not hard to implement. I'll do it after #1490 will be merged.

**Comment by popzxc** [comment]
As @alekseysidorov requested, I also added a possibility to add builtin instance collections.

So users of `exonum-cli` can deploy&init service at the start of the Exonum node.

File: cli/src/lib.rs [hunk 4]
```diff
@@ -130,6 +132,26 @@ impl NodeBuilder {
...
+    }
+
+    /// Adds a new InstanceCollection to the list of available services.
+    ///
+    /// With that method you can start Exonum with already deployed & running service
+    /// instances.
+    ///
+    /// Note that if you add the `InstanceCollection` manually, you don't need to call
+    /// `with_service`, since `InstanceCollection` already contains a service inside.
+    pub fn with_instance_collection(mut self, instance_collection: InstanceCollection) -> Self {
```
Comment (alekseysidorov) [inline cli/src/lib.rs]: I mean that we should provide runtime-agnostic mechanism to add instances at startup.

Something like this:

```rust
struct Instance {
    name: String,
    artifact: ArtifactId,
    constructor: Vec<u8>
}
```

Comment (popzxc) [reply cli/src/lib.rs]: Ah. I see. I thought you need it as a temporary solution to run services without `exonum-launcher`.

I'll remove it then.

Comment (popzxc) [reply cli/src/lib.rs]: However, runtime-independent deploy at the start is much bigger question, and totally out of the scope of this PR.

**Pull request merged**